src/thingset.h

```diff
@@ -529,7 +533,16 @@ void ts_dump_json(struct ts_context *ts, ts_object_id_t obj_id, int level);
  * @param ts Pointer to ThingSet context.
  * @param flags Flags to define authentication level (1 = access allowed)
  */
-void ts_set_authentication(struct ts_context *ts, uint16_t flags);
+void ts_set_authentication(struct ts_context *ts, uint8_t flags);
+
+/**
+ * Checks if any of the data objects belonging to the subsets was updated.
+ *
+ * @param ts Pointer to ThingSet context.
+ * @param subsets Flags to select which subset(s) of data items should be considered
+ * @param clear If set to true, the updated status is cleared after checking.
+ */
+bool ts_check_updated(struct ts_context *ts, const uint16_t subsets, bool clear);
```

Comment:
Do we really need that? Isn't there anyway a change callback that may be used by an application?

Reply:
> Maybe a better approach would be to assign a callback for every patch request instead. I'll think about this, thanks for the pointer.

A ThingSet context may have an update counter. On every patch request the update counter may be incremented and every data object that is changed will have it's own update counter set to the context's update count. The general callback can then search for updated data objects by the update count. As the update count in the data objects should be of small size (e.g. 2 bits) all object update counts that are equal to the current one shall be invalidated before a patch request is applied.  By this the last (e.g. 4) "transactions" on the object database can be retrieved by an application. This may help for multi patch request updates to the object database.

The general patch callback should have a void(*cb)(uintxx_t update_count, other???) signature to care for update count races in the execution of several patch requests (the context update count may already be incremented again when the callback is executed).

Reply:
> Maybe a better approach would be to assign a callback for every patch request instead. I'll think about this, thanks for the pointer.

A ThingSet context may have an update counter. On every patch request the update counter may be incremented and every data object that is changed will have it's own update counter set to the context's update count. The general callback can then search for updated data objects by the update count. As the update count in the data objects should be of small size (e.g. 2 bits) all object update counts that are equal to the current one shall be invalidated before a patch request is applied. By this the last (e.g. 4) "transactions" on the object database can be retrieved by an application. This may help for multi patch request updates to the object database.

The general patch callback should have a void(*cb)(uintxx_t update_count, other???) signature to care for update count races in the execution of several patch requests (the context update count may already be incremented again when the callback is executed).

Note: Commented again - last comment got somehow lost in GitHub

Reply:
I'm not sure if I understand correctly. How would the counter value be transferred over the wire?

I don't think we should support multiple patch requests in parallel. If one patch request is still in progress, the next requester will have to wait until the request is finished.

Reply:
> I'm not sure if I understand correctly. How would the counter value be transferred over the wire?

Never !-) The counter is just a way to mark a data object was changed. This way the callback function can search the database for changed data objects. The counter just helps to in case of asynchronous callback function execution and a new patch executed at the same time the callback function is still active from the last patch request. If we serialise patch -> callback -> patch a data changed flag would be sufficient. 

> I don't think we should support multiple patch requests in parallel. If one patch request is still in progress, the next requester will have to wait until the request is finished.

Yes - patch requests to the database have to be serialised. The design decision is whether a patch callback that is executed after the patch is applied can block further patch request to be executed until it finishes. In my opinion this should not be the case.

Reply:
I have reverted the ts_check_updated function and introduced a callback for the patch functions. This avoids to iterate through all data objects for each patch request and is quite similar to the previous approach where the callback was assigned per group.

Now the application can decide what to do after a patch request. It can either store the updated data in NVM synchronously in the callback (which would delay the ThingSet response) or it stores the data asynchronously and takes care of any race conditions. Handling asynchronous data processing inside the library will make it too complicated for very simple devices with low processing power, I think.